pkg/controller/core/resource_controller.go

```diff
@@ -47,17 +47,41 @@ const (
 	errorSettingResourceBindStatus   = "Failed to set resource binding status"
 	errorResettingResourceBindStatus = "Failed to reset resource binding status"
 	waitResourceIsNotAvailable       = "Waiting for resource to become available"
-
-	// RequeueOnWait - requeue after duration when waiting for resource state
-	RequeueOnWait = 30 * time.Second
-	// RequeueOnSuccess - requeue after duration when resources status is successful
-	RequeueOnSuccess = 2 * time.Minute
 )
 
-// Commonly used reconciliation results.
+// Suggested reconcile results. A reconciliation that either altered the
+// requested Kubernetes object or returned an error is considered dirty.
+// Reconciliations that do not alter the object and do not return an error are
+// considered clean.
 var (
-	Result        = reconcile.Result{}
-	ResultRequeue = reconcile.Result{Requeue: true}
+	// RequeueIfDirty does not queue a new reconciliation. A reconciliation will
+	// be queued implicitly if the current reconciliation was dirty. Prefer
+	// RequeueIfDirty if either the current reconciliation is known to be dirty,
+	// or further reconciliation is not necessary until the requested object is
+	// altered outside of this controller.
+	RequeueIfDirty = reconcile.Result{Requeue: false}
```

Comment:
I am not :100: sure about that. How does the developer know if the object has been modified, i.e. "is dirty"?
For example
```go
...
	b := &v1alpha1.Bucket{}
	if err := r.Get(ctx, request.NamespacedName, b); err != nil {
		if kerrors.IsNotFound(err) {
			return reconcile.Result{}, nil
		}
		return reconcile.Result{}, err
	}
// some blob of code here followed by
  return RequeueIfDirty, r.Status().Update(ctx, b)
}
```
The `controller-runtime` does a good job by tracking object state ("dirty" or otherwise), i.e. it absolves me from this task. Introducing `RequeueIfDirty` means that now I must be aware of the object state changes and chose an appropriate `Requeue` result. My initial and current feeling about it - we don't need this one.

Reply:
> How does the developer know if the object has been (or will be) modified, i.e. "is dirty", or what "Dirty" is in the first plane and what are the implications of it?

Quoting the Godoc I added:

> A reconciliation that either altered the requested Kubernetes object or returned an error is considered dirty. Reconciliations that do not alter the object and do not return an error are considered clean.

Note that I'm considering the entire result of of the reconcile as 'clean' or 'dirty'. i.e. The reconcile is dirty if the object didn't change but did return an error.

The key here is Requeue __If__ Dirty. It's trying to document the not-very-obvious fact that the controller _might_ requeue anyway in these scenarios. I don't expect the developer to check whether the object changed or not.

Reply:
`RequeueIfDirty` is attempting to be a succinct way of saying "I'm not asking you to requeue explicitly, but I understand that the controller might requeue anyhow". Thus allowing us to avoid having 10 different names for the same `reconcile.Result{Requeue: false}`.

Reply:
I see, thank you for an explanation. With that said:
`RequeueIfDirty` vs. `reconcile.Result{}` (or `reconciler.Result{Requeue: false}`).
In my opinion, the latter two a much more intuitive, especially for a newcomer, who may not know what the "Dirty" even means, while with the `result` he/she explicitly say: "no, do not requeue (please)" :)